backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/domain/CourseExecution.java

```diff
@@ -222,4 +236,17 @@ public int getNumberOfQuestions() {
                 .filter(user -> user.getRole().equals(User.Role.STUDENT))
                 .collect(Collectors.toSet());
     }
+
+    public Set<Topic> findAvailableTopics() {
+        return getAvailableAssessments().stream().flatMap(assessment -> assessment.getTopics().stream()).distinct().collect(Collectors.toSet());
+    }
+
+    public List<Question> filterQuestionsByTopics(List<Question> questions, Set<TopicDto> topics) {
+        List<Integer> availableTopicsIds = findAvailableTopics().stream().map(Topic::getId).collect(Collectors.toList());
+        List<Integer> topicsIds = topics.stream().map(TopicDto::getId).distinct().collect(Collectors.toList());
+
```

Comment:
it can be further simplified if we only use the topics that are common to both collections
Set<Integer> availableTopicsIds = findAvailableTopics().stream().map(Topic::getId).collect(Collectors.toSet());
Set<Integer> topicsIds = topics.stream().map(TopicDto::getId).filter(topicId -> availableTopicIds.contains(topicId)).collectCollectors.toSet());

Reply:
Done